### PR TITLE
Re-enable BatchNorm autodiff

### PR DESCRIPTION
**Pull request by jjsjann123**
Turns on BN in autodiff:

1. outputs an empty tensor for running stats to by pass autodiff issue on None;
2. fixing BN inference backward in cudnn & miopen, where backward falls back to native batchnorm kernel instead;

### CODE REVIEW

**Comment by jjsjann123** [comment]
Do we have a batchnorm implemented in xla?
I'm updating the empty tensor output for BN in pytorch to make it JIT friendly, but can't find the XLA backend code to make it consistent.
Tagging @ailzhang in case you happen to know it.

```
May 01 00:41:20 [ RUN      ] TensorTest.TestBatchNorm1D
May 01 00:41:20 /var/lib/jenkins/workspace/xla/test/cpp/test_tensor.cpp:417: Failure
May 01 00:41:20 Expected equality of these values:
May 01 00:41:20   std::get<1>(output).defined()
May 01 00:41:20     Which is: true
May 01 00:41:20   !std::get<1>(xla_output).is_null()
May 01 00:41:20     Which is: false
May 01 00:41:20 /var/lib/jenkins/workspace/xla/test/cpp/test_tensor.cpp:419: Failure
May 01 00:41:20 Expected equality of these values:
May 01 00:41:20   std::get<2>(output).defined()
May 01 00:41:20     Which is: true
May 01 00:41:20   !std::get<2>(xla_output).is_null()
May 01 00:41:20     Which is: false
May 01 00:41:20 /var/lib/jenkins/workspace/xla/test/cpp/test_tensor.cpp:417: Failure
May 01 00:41:20 Expected equality of these values:
May 01 00:41:20   std::get<1>(output).defined()
May 01 00:41:20     Which is: true
May 01 00:41:20   !std::get<1>(xla_output).is_null()
May 01 00:41:20     Which is: false
May 01 00:41:20 /var/lib/jenkins/workspace/xla/test/cpp/test_tensor.cpp:419: Failure
May 01 00:41:20 Expected equality of these values:
May 01 00:41:20   std::get<2>(output).defined()
May 01 00:41:20     Which is: true
May 01 00:41:20   !std::get<2>(xla_output).is_null()
May 01 00:41:20     Which is: false
```

**Comment by jjsjann123** [comment]
The rocm failure looks real but doesn't make sense to me.

From the log, looks like `BatchNorm1d_3d_input_eval_cuda` runs fine, but `BatchNorm1d_affine_eval_cuda` gets a null address access?

The only change in miopen are these: https://github.com/pytorch/pytorch/blob/66b013fc13eedd56680bf59941fe83eb26aad3a0/aten/src/ATen/native/miopen/BatchNorm_miopen.cpp#L123-L124, which should really affect the two test cases the same way.

```
01:00:47   test_BatchNorm1d_3d_input (__main__.TestNN) ... [W Module.cpp:514] Warning: Disabling benchmark mode for MIOpen is NOT supported. Overriding value to True (function operator())
01:00:48 ok (0.595s)
01:00:48   test_BatchNorm1d_3d_input_cuda (__main__.TestNN) ... ok (0.577s)
01:00:48   test_BatchNorm1d_3d_input_eval (__main__.TestNN) ... ok (0.029s)
01:00:49   test_BatchNorm1d_3d_input_eval_cuda (__main__.TestNN) ... ok (0.118s)
01:00:49   test_BatchNorm1d_3d_input_not_affine (__main__.TestNN) ... ok (0.041s)
01:00:49   test_BatchNorm1d_3d_input_not_affine_cuda (__main__.TestNN) ... ok (0.116s)
01:00:49   test_BatchNorm1d_3d_input_not_affine_eval (__main__.TestNN) ... ok (0.017s)
01:00:49   test_BatchNorm1d_3d_input_not_affine_eval_cuda (__main__.TestNN) ... ok (0.110s)
01:00:49   test_BatchNorm1d_affine (__main__.TestNN) ... ok (0.499s)
01:00:49   test_BatchNorm1d_affine_cuda (__main__.TestNN) ... ok (0.674s)
01:00:50   test_BatchNorm1d_affine_eval (__main__.TestNN) ... ok (0.475s)
01:00:51   test_BatchNorm1d_affine_eval_cuda (__main__.TestNN) ... Memory exception on virtual address 0x0, node id 4 : Page not present
01:00:51 Address does not belong to a known buffer
01:00:51 Memory access fault by GPU node-4 (Agent handle: 0x5568318a7ac0) on address (nil). Reason: Page not present or supervisor privilege.
01:00:56 Traceback (most recent call last):
01:00:56   File "test/run_test.py", line 1172, in <module>
01:00:56     main()
01:00:56   File "test/run_test.py", line 1151, in main
01:00:56     raise RuntimeError(err_message)
01:00:56 RuntimeError: test_nn failed! Received signal: SIGIOT
```

https://github.com/pytorch/pytorch/blob/66b013fc13eedd56680bf59941fe83eb26aad3a0/torch/testing/_internal/common_nn.py#L1307-L1315
https://github.com/pytorch/pytorch/blob/66b013fc13eedd56680bf59941fe83eb26aad3a0/torch/testing/_internal/common_nn.py#L1316-L1324

**Comment by jjsjann123** [comment]
I think I got the XLA test case covered here: https://github.com/pytorch/xla/pull/2961 (at least the test case passes locally, :crossed_fingers: )

Anyway I can test the rocm failure? @ngimel

**Comment by jjsjann123** [comment]
Note to myself. With the update in BN native kernel using TensorIterators, the change in this PR doesn't work any more.
I'll update it soon.

**Comment by jjsjann123** [comment]
rocm failure seems to be not related. Please let me know if I missed anything important there.

File: test/test_jit.py
Comment (Krovatkin) [inline test/test_jit.py:10567]: similar questions to the layernorm:

Should we confirm that batchnorm was indeed auto differentiated?
Also
this test looks very similar to the tests jit_metaprogramming_utils.py should be doing:
fwd result check against ref
grad check against grad_ref
grad grad check against grad grad ref
Is there a reason for a separate test?

File: aten/src/ATen/native/cuda/Normalization.cu
Comment (albanD) [inline aten/src/ATen/native/cuda/Normalization.cu:464]: Given the modification on the caller, should this assert that save_mean is always defined?
Same for save_invstd below

File: aten/src/ATen/native/Normalization.cpp
Comment (albanD) [inline aten/src/ATen/native/Normalization.cpp:465]: Is this TODO for later?

File: aten/src/ATen/native/Normalization.cpp
Comment (albanD) [inline aten/src/ATen/native/Normalization.cpp:464]: Was this failing before this PR?
If so, could we add a test in test_nn.py that makes sure that we don't introduce this again in the future.

File: test/test_jit.py
Comment (jjsjann123) [reply test/test_jit.py:10567]: Good catch, I'm copying the response in the other PR: https://github.com/pytorch/pytorch/pull/50467#discussion_r673040780

File: aten/src/ATen/native/Normalization.cpp
Comment (jjsjann123) [reply aten/src/ATen/native/Normalization.cpp:465]: Yes. I would need confirmation from someone who know rocm and its libraries to verify that.

File: aten/src/ATen/native/Normalization.cpp
Comment (jjsjann123) [reply aten/src/ATen/native/Normalization.cpp:464]: I vaguely remember that we run into a failing case in the test added below. Nick has mentioned about moving that test into meta test set: https://github.com/pytorch/pytorch/pull/57321/files/2855c79bbd43fc2135fc238b96d3a19173b5567b#r670053319

File: aten/src/ATen/native/cuda/Normalization.cu
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/Normalization.cu:464]: I think so. I can try adding that and run it through CI to see if we upset anything.

File: aten/src/ATen/native/Normalization.cpp
Comment (albanD) [reply aten/src/ATen/native/Normalization.cpp:464]: Ok, we should make sure we have a test for it in this PR.

File: test/test_jit.py
Comment (jjsjann123) [reply test/test_jit.py:10757]: I have pushed the test to meta programming set.
However, I don't see how to specify test to run on GPU in the meta programming. So I have to kept this test to verify the fix for inference backward in cudnn.

File: aten/src/ATen/native/Normalization.cpp
Comment (jjsjann123) [reply aten/src/ATen/native/Normalization.cpp:515]: I verified the failure and fix in the updated python test.

**Comment by jjsjann123** [comment]
Addressed review comments. good for a second review while waiting on CI :crossed_fingers:

**Comment by jjsjann123** [comment]
Oops, looks like a real failure on added test, I'm looking into this.

**Comment by jjsjann123** [comment]
Didn't spot any thing in failing CI.
I think we are good for review cc'ing @albanD @Krovatkin

File: aten/src/ATen/native/Normalization.cpp
Comment (albanD) [inline aten/src/ATen/native/Normalization.cpp:244]: Why do you need to do that?
Also this might not be contiguous anymore?

Comment (jjsjann123) [reply aten/src/ATen/native/Normalization.cpp:244]: `weight` could be undefined. We run into issue here with BN with only bias. e.g.

https://github.com/pytorch/pytorch/pull/57321/files#diff-4a85700059a0d89291a89a82d2f1ce111293a80c5bc7009b0479fb51b7e5d548R119-R121

> Also this might not be contiguous anymore?
We are creating fresh output 1d buffer, I don't quite get which is non-contiguous in this context.

**Comment by jjsjann123** [comment]
Kindly pinging @albanD

File: aten/src/ATen/native/Normalization.cpp
Comment (albanD) [reply aten/src/ATen/native/Normalization.cpp:244]: cc @zou3519 could this be breaking some existing batched Tensor code?
How do we deal with this case where weight might be undefined, should we just check and call different functions?

Comment (zou3519) [reply aten/src/ATen/native/Normalization.cpp:244]: There is no problem for existing BatchedTensor code because native_batch_norm is a primitive w.r.t. autograd (it has a backwards formula) BatchedTensor never sees this code.

**Comment by csarofeen** [comment]
Hey, @albanD is there anything else you think should be addressed in this PR?

File: aten/src/ATen/native/cuda/Normalization.cu
Comment (albanD) [inline aten/src/ATen/native/cuda/Normalization.cu:484]: Isn't it possible that the save_mean has 0 elements even in the case where you're not training?

File: aten/src/ATen/native/Normalization.cpp
Comment (albanD) [reply aten/src/ATen/native/Normalization.cpp:512]: I don't see any update to `test_nn` here corresponding to that?

**Review by albanD — COMMENTED** [review]
Sorry for the delay in review. But this makes some subtle changes to a pretty subtle and undocumented API so I am not super confident I grasp all the repercutions.

**Comment by csarofeen** [comment]
Understood, just trying hard to do whatever we can to get this in so we can finally start pushing nvFuser. Thanks, pinging back to @jjsjann123 and @rdspring1.

File: aten/src/ATen/native/Normalization.cpp
Comment (jjsjann123) [reply aten/src/ATen/native/Normalization.cpp:512]: this is covered in https://github.com/pytorch/pytorch/pull/57321/files#diff-c54c3bae2ea206c4de5b90adf7f2d3ec47df6818451b8dd2b3c36fe54adc10d8R10759

That's probably not a good place for the purpose of the test. I'll move it to `test_nn.py` with a simplified version, since the majority of the old test is covered in generated jit metaprogramming test.

File: aten/src/ATen/native/cuda/Normalization.cu
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/Normalization.cu:484]: So for inference, when normalizing with running_mean and running_var, we are not computing save_mean.
for native_batch_norm, we are copying running_mean to save_mean, even though we are not computing them; while for cudnn_batch_norm, save_mean is left as 0-sized tensor.

File: aten/src/ATen/native/Normalization.cpp
Comment (jjsjann123) [reply aten/src/ATen/native/Normalization.cpp:512]: Sorry, I was wrong earlier. I don't think this fix applies to test_nn.

Autograd has already handled this, when we run with inference, we fallback to native_batch_norm_backward. https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L2118

This is exactly what we are doing here for autodiff. And I do think `test_jit.py` is the right place to verify that fix.

File: aten/src/ATen/native/cuda/Normalization.cu
Comment (albanD) [reply aten/src/ATen/native/cuda/Normalization.cu:484]: So the answer is yes?
In particular, what if the input has a dimension of size 0 or is a scalar?

Comment (jjsjann123) [reply aten/src/ATen/native/cuda/Normalization.cu:484]: Good catch.
I think we would need to duplicate this short-cut into `_batch_norm_impl_index(_backward)` since backends doesn't really support size-0 inputs.

Comment (jjsjann123) [reply aten/src/ATen/native/cuda/Normalization.cu:484]: https://github.com/pytorch/pytorch/blob/000e0689816596fada756234dbc4cf0a73790876/aten/src/ATen/native/Normalization.cpp#L534-L540

Comment (jjsjann123) [reply aten/src/ATen/native/cuda/Normalization.cu:491]: fixed in this: https://github.com/pytorch/pytorch/pull/57321/commits/7222cb3d28f9a25b280ded6c0031ff45f5b8e324

**Review by albanD — APPROVED** [review]
Okay

**Comment by albanD** [comment]
@jansel can you take a final look and land this if you're happy with it?

**Comment by jansel** [comment]
> @jansel can you take a final look and land this if you're happy with it?

Will do, thanks @albanD!

**Comment by jansel** [comment]
@jjsjann123 there are some FB internal lints from clang-tidy that prevent landing:
```
Summary: 
caffe2/aten/src/ATen/native/Normalization.cpp:243:10: error: unused variable 'options' [-Werror,-Wunused-variable]
    auto options = input.options();
         ^
stderr: caffe2/aten/src/ATen/native/Normalization.cpp:243:10: error: unused variable 'options' [-Werror,-Wunused-variable]
    auto options = input.options();
```

Mind fixing that?  (Looks like it gets recomputed in the next line.)

File: aten/src/ATen/native/Normalization.cpp
Comment (ngimel) [inline aten/src/ATen/native/Normalization.cpp:541]: I don't think this actually works, as `num_batch` is likely 0, you'll get 
```
In [33]: a=torch.randn(0,2,3,4)

In [34]: a.view(0,2,-1)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-34-845acd0bbe9f> in <module>
----> 1 a.view(0,2,-1)

RuntimeError: cannot reshape tensor of 0 elements into shape [0, 2, -1] because the unspecified dimension size -1 can be any value and is ambiguous

```
It also means that this is likely untested.

File: aten/src/ATen/native/Normalization.cpp
Comment (ngimel) [inline aten/src/ATen/native/Normalization.cpp:534]: why do you need this clone? (It's cheap for an empty tensor of course, but still)

File: aten/src/ATen/native/Normalization.cpp
Comment (ngimel) [inline aten/src/ATen/native/Normalization.cpp:421]: can you please leave a comment that under normal conditions this should never happen, and you will come here only from symbolic script.
Also, to delete this once you no longer rely on symbolic script and switch to lazy tensor.

File: aten/src/ATen/native/Normalization.cpp
Comment (jjsjann123) [reply aten/src/ATen/native/Normalization.cpp:541]: Oops, good catch.
I put 0-dim in the wrong spot in my test :facepalm: . https://github.com/pytorch/pytorch/pull/57321/files#diff-4a85700059a0d89291a89a82d2f1ce111293a80c5bc7009b0479fb51b7e5d548R115-R120

Will update.

File: aten/src/ATen/native/Normalization.cpp
Comment (jjsjann123) [reply aten/src/ATen/native/Normalization.cpp:534]: This was blindly copied from `batch_norm`. 
https://github.com/pytorch/pytorch/pull/57321/files#diff-9e36531b6ea57776251e04cf2e1b84a94145020d870fed3d7f82388c47371483L533

I remember we had issue with `view`, though I think that's only with autodiff, but I was too scared to touch it.
Let me try to change it and see if the world blows up.

File: aten/src/ATen/native/Normalization.cpp
Comment (jjsjann123) [reply aten/src/ATen/native/Normalization.cpp:421]: I had a long conversation on this topic with Alban offline.

We don't want exposed logic that would ended up with control flow in autodiff, because it breaks fusion :(, that defeats all the good things with supporting BN in autodiff.

This is not only for symbolic script. In eager mode you would also come here (we moved the short cut from `batch_norm` here, because we don't want to duplicate it). Arguably we could put that comment in `_batch_norm_impl_index_backward`, but in face you'll only go to that function through `symbolic_script`, so that seems to be a redundant comment.

Comment (jjsjann123) [reply aten/src/ATen/native/Normalization.cpp:421]: Looking at the stripped `batch_norm`, I think it makes sense to update the API so we no longer need `_batch_norm_impl_index`.
This makes `symbolic_script.cpp` more straightforward, so that we don't have to rewrite `batch_norm` with a different operation `_batch_norm_impl_index`, which feels pretty awkward.

I know updating the API seems to be a big hammer. Any comment on that?

File: aten/src/ATen/native/Normalization.cpp
Comment (jjsjann123) [reply aten/src/ATen/native/Normalization.cpp:541]: Double oops, I also forgot to put weight/bias on those test cases :facepalm: 
I have no idea what I was doing....

File: aten/src/ATen/native/Normalization.cpp
Comment (jjsjann123) [reply aten/src/ATen/native/Normalization.cpp:534]: took another look at this, won't we forwarding `input` with the absence of weight/bias. That's going to be surprising and tricky to debug. I think we should keep it as is.

File: aten/src/ATen/native/Normalization.cpp
Comment (albanD) [inline aten/src/ATen/native/Normalization.cpp:424]: note than in eager mode, this creates (potentially large) Tensors that just get discarded on exit....

File: aten/src/ATen/native/Normalization.cpp
Comment (albanD) [reply aten/src/ATen/native/Normalization.cpp]: `_batch_norm_impl_index_backward` is a native function with some alias informations associated with it. So we have to respect it.
 In particular, that function declare that the computed gradient doesn't alias the grad_output. So it shouldn't

File: aten/src/ATen/native/Normalization.cpp
Comment (ngimel) [inline aten/src/ATen/native/Normalization.cpp:530]: cool, where is a test that should have been failing previously?

Comment (jjsjann123) [reply aten/src/ATen/native/Normalization.cpp:530]: https://github.com/pytorch/pytorch/pull/57321/files#diff-4a85700059a0d89291a89a82d2f1ce111293a80c5bc7009b0479fb51b7e5d548R115-R122

I verified the failure before applying the fix.

File: aten/src/ATen/native/Normalization.cpp
Comment (jjsjann123) [reply aten/src/ATen/native/Normalization.cpp:424]: Yeah, that's unfortunate.
But if we are to stick to our plan of unifying the APIs, this will happen to even wider extend... I'll open an upstream issue so we can have a discussion there!

**Review by ngimel — APPROVED** [review]
This looks ok to me, but overall that's a lot of code complexity to accommodate symbolic script. Longer term we should either think about how to refactor the code so that's not necessary, or not rely on symbolic script to get backward graph (use lazy tensor? Python dispatch?)

**Pull request closed without merging**
